components/action-mgt/org.wso2.carbon.identity.action.mgt/src/main/java/org/wso2/carbon/identity/action/mgt/dao/impl/ActionManagementDAOImpl.java

```diff
@@ -0,0 +1,389 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.action.mgt.dao.impl;
+
+import org.wso2.carbon.identity.action.mgt.constant.ActionMgtConstants;
+import org.wso2.carbon.identity.action.mgt.constant.ActionMgtSQLConstants;
+import org.wso2.carbon.identity.action.mgt.dao.ActionManagementDAO;
+import org.wso2.carbon.identity.action.mgt.exception.ActionMgtException;
+import org.wso2.carbon.identity.action.mgt.exception.ActionMgtServerException;
+import org.wso2.carbon.identity.action.mgt.model.Action;
+import org.wso2.carbon.identity.action.mgt.model.AuthType;
+import org.wso2.carbon.identity.action.mgt.model.EndpointConfig;
+import org.wso2.carbon.identity.action.mgt.util.ActionManagementUtil;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static org.wso2.carbon.identity.action.mgt.constant.ActionMgtConstants.ACTIVE_STATUS;
+import static org.wso2.carbon.identity.action.mgt.constant.ActionMgtConstants.INACTIVE_STATUS;
+
+/**
+ * This class implements the {@link ActionManagementDAO} interface.
+ */
+public class ActionManagementDAOImpl implements ActionManagementDAO {
+
+    @Override
+    public Action addAction(String actionType, Action action, Integer tenantId)
+            throws ActionMgtException {
+
+        String generatedActionId = UUID.randomUUID().toString();
+        try (Connection dbConnection = IdentityDatabaseUtil.getDBConnection(true)) {
+            try (PreparedStatement prepStmt = dbConnection.prepareStatement(
+                        ActionMgtSQLConstants.ADD_ACTION_TO_ACTION_TYPE)) {
+                prepStmt.setString(1, generatedActionId);
+                prepStmt.setString(2, actionType);
+                prepStmt.setString(3, action.getName());
+                prepStmt.setString(4, action.getDescription());
+                prepStmt.setString(5, ACTIVE_STATUS);
+                prepStmt.setInt(6, tenantId);
+                prepStmt.executeUpdate();
+                prepStmt.clearParameters();
+
+                // Add Endpoint configuration properties.
+                addEndpointProperties(dbConnection, generatedActionId, getEndpointProperties(action),
+                        tenantId);
+
+                IdentityDatabaseUtil.commitTransaction(dbConnection);
+
+                return getActionByActionId(generatedActionId, tenantId);
+            } catch (SQLException | ActionMgtException e) {
+                IdentityDatabaseUtil.rollbackTransaction(dbConnection);
+                throw e;
```

Comment:
Is this been a practice to throw the exception at rollback as it is ?
Shouldn't this be thrown as a server error

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/5771/commits/94e90e62016b434a0f37aa2b1f7ff0edc34bb64f